rotkehlchen/chain/evm/decoding/curve/curve_cache.py

```diff
@@ -0,0 +1,475 @@
+import logging
+from typing import TYPE_CHECKING, Literal, NamedTuple
+
+from rotkehlchen.assets.asset import UnderlyingToken
+from rotkehlchen.assets.utils import TokenEncounterInfo, get_or_create_evm_token
+from rotkehlchen.chain.evm.constants import ETH_SPECIAL_ADDRESS, ZERO_ADDRESS
+from rotkehlchen.chain.evm.contracts import EvmContract
+from rotkehlchen.chain.evm.decoding.curve.constants import (
+    CURVE_ADDRESS_PROVIDER,
+    CURVE_API_URLS,
+    CURVE_METAREGISTRY_METHODS,
+    IGNORED_CURVE_POOLS,
+)
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.constants import ONE
+from rotkehlchen.db.addressbook import DBAddressbook
+from rotkehlchen.errors.misc import (
+    InputError,
+    NotERC20Conformant,
+    RemoteError,
+    UnableToDecryptRemoteData,
+)
+from rotkehlchen.errors.serialization import DeserializationError
+from rotkehlchen.globaldb.cache import (
+    globaldb_get_general_cache_like,
+    globaldb_get_general_cache_values,
+    globaldb_get_unique_cache_value,
+    globaldb_set_general_cache_values,
+    globaldb_set_unique_cache_value,
+    read_curve_pool_tokens,
+)
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_evm_address
+from rotkehlchen.types import (
+    CURVE_POOL_PROTOCOL,
+    AddressbookEntry,
+    CacheType,
+    ChainID,
+    ChecksumEvmAddress,
+    EvmTokenKind,
+)
+from rotkehlchen.utils.network import request_get_dict
+
+if TYPE_CHECKING:
+    from gevent import DBCursor
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class CurvePoolData(NamedTuple):
+    pool_address: ChecksumEvmAddress
+    pool_name: str
+    lp_token_address: ChecksumEvmAddress
+    gauge_address: ChecksumEvmAddress | None
+    coins: list[ChecksumEvmAddress]
+    underlying_coins: list[ChecksumEvmAddress] | None
+
+
+class CurveCache:
+    chain_id: ChainID
+
+    def __init__(self: 'CurveCache', chain_id: ChainID) -> None:
+        self.chain_id = chain_id
+
+    def read_curve_pools_and_gauges(self: 'CurveCache') -> tuple[dict[ChecksumEvmAddress, list[ChecksumEvmAddress]], set[ChecksumEvmAddress]]:  # noqa: E501
+        """Reads globaldb cache and returns:
+        - A set of all known curve pools addresses for the given chain.
+        - A set of all known curve gauges addresses for the given chain.
+
+        Returned functions doesn't raise anything unless cache entries were inserted incorrectly.
+        """
+        with GlobalDBHandler().conn.read_ctx() as cursor:
+            curve_pools_lp_tokens = globaldb_get_general_cache_values(
+                cursor=cursor,
+                key_parts=(CacheType.CURVE_LP_TOKENS, str(self.chain_id.serialize_for_db())),
```

Comment:
Instead of calling multiple times `str(self.chain_id.serialize_for_db())` do it once and save it in a variable